core/templates/pages/exploration-player-page/services/question-player-engine.service.spec.ts

```diff
@@ -0,0 +1,1255 @@
+  });
+
+  it('should update the exploration version if it' +
+    'is not in question player mode when initialized', fakeAsync(() => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+    let exploration = sampleExplorationDict;
+    exploration.version = 2;
+
+    spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+    spyOn(contextService, 'getExplorationId').and.returnValue('explorationId1');
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(false);
+    spyOn(urlService, 'getExplorationVersionFromUrl').and.returnValue(1);
+
+    let readOnlySpy =
+      spyOn(readOnlyExplorationBackendApiService, 'loadExplorationAsync')
+        .and.callFake((expID, version) => {
+          expect(version).toBe(1);
+          return Promise.resolve(exploration);
+        });
+
+    questionPlayerEngineService.init(
+      multipleQuestionBackendDict, successHandler, failHandler);
+    tick();
+
+    expect(questionPlayerEngineService.questionPlayerMode).toBe(false);
+    expect(readOnlySpy).toHaveBeenCalled();
+    expect(questionPlayerEngineService.getExplorationVersion()).toBe(2);
+  }));
+
+  it('should return the current question', () => {
+    let successCallback = jasmine.createSpy('success');
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+    let exploration = sampleExplorationDict;
+    let answer = 'answer';
+    let interactionRulesService: InteractionRulesService;
+    let answerClassificationResult: AnswerClassificationResult = {
+      answerGroupIndex: 0,
+      classificationCategorization: 'default_outcome',
+      outcome: {
+        dest: null,
+        feedback: {
+          _contentId: 'feedback_id',
+          _html: '<p>Dummy Feedback</p>',
+          contentId: 'feedback_id',
+          html: '<p>Dummy Feedback</p>',
+        },
+        labelledAsCorrect: true,
+        missingPrerequisiteSkillId: null,
+        paramChanges: [],
+        refresherExplorationId: null,
+      },
+      ruleIndex: 0
+    } as unknown as AnswerClassificationResult;
+
+    spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+    spyOn(contextService, 'getExplorationId').and.returnValue('explorationId1');
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(true);
+    spyOn(urlService, 'getExplorationVersionFromUrl').and.returnValue(1);
+    spyOn(readOnlyExplorationBackendApiService, 'loadExplorationAsync')
+      .and.returnValue(Promise.resolve(exploration));
+    let answerClassificationServiceSpy =
+      spyOn(answerClassificationService, 'getMatchingClassificationResult')
+        .and.returnValue(answerClassificationResult);
+    spyOn(expressionInterpolationService, 'processHtml')
+      .and.callFake((html, envs) => {
+        if (html === null) {
+          return null;
+        }
+        return html;
```

Comment:
Isn't this equivalent to "return html" (in all cases)?

Reply:
No, in some specific tests we are testing a warning message when the received element is null.

Reply:
Let me rephrase. From my reading of the code, can't this entire inner function be replaced with "return html"?

Reply:
My bad :D